### ISSUE TITLE
HasPlatformType should be more strict

### ISSUE THREAD

**Issue by BraisGabin** [detekt/detekt#3603]
## Expected Behavior of the rule

```kotlin
val path = URLEncoder().encode("urlPath")
```

This code should be flagged because `path` have a platform type

this code should be:

```kotlin
val path = URLEncoder().encode("urlPath")!!
```
or
```kotlin
val path: String = URLEncoder().encode("urlPath")
```
or
```kotlin
val path: String? = URLEncoder().encode("urlPath")
```

## Context

#3571

Other code that should be flagged:

```kotlin
"${URLEncoder().encode("urlPath")}"
```

```kotlin
URLEncoder().encode("urlPath").toUpperCase()
```

```kotlin
hello(URLEncoder().encode("urlPath"))
```

**Comment by severn-everett** [comment]
A couple of points:
- Which `URLEncoder` is being referred to here? The `URLEncoder` class from the `java.net` package doesn't have a public constructor.
- I'm not sure that there should be a check on string interpolation in and of itself. String interpolation can accept null values, so it wouldn't matter the `URLEncoder.encode()` function returned null in `"${URLEncoder.encode("urlPath")}"` - it'd just print out ``. It'd not be ideal, but it wouldn't be something that would necessary raise an NPE.

**Comment by BraisGabin** [comment]
> it'd just print out a blank string

That is not true. It would print `"null"`. For that reason we have `NullableToStringCall`. And that rule doesn't want to mess with Platform Types. For that reason I think that this rule should handle them.

**Comment by severn-everett** [comment]
Are all of these new checks only for public function and property declarations like how the rule is currently configured? Otherwise, you'd get situations where assigning a local property a flexible type would be fine, but then using that flexible local property in a chain would be flagged.

**Comment by BraisGabin** [comment]
Nope, it applies to any code that returns an felxible type. You should assert about it's nullabilityat that moment. Even if you are assigning a local property. The local property shouldn't have a flexible type.

**Comment by severn-everett** [comment]
Ok. A warning that this is going to affect a _lot_ of code in our own codebase; I'll go into more detail about it in the PR when the next commit gets pushed.

**Comment by BraisGabin** [comment]
That's good. Then I'll talk about disable `UnsafeCallOnNullableType` on our code base. Or update it so the usage of `!!` for platform types could be allowed (adding a new configuration).

**Comment by fwh-dc** [comment]
What's the status of this? I think this feature would be extremely useful.

**Comment by cortinico** [comment]
> What's the status of this? I think this feature would be extremely useful.

This is up for grab. Do you want to take a stance at it @fwh-dc ?

**Comment by fwh-dc** [comment]
I could try it out. I'd like to know:

How do I run the test cases for this particular rule?
And is there some documentation on the internals of this module or should I just look around and see how other rules are implemented?

:)

**Comment by BraisGabin** [comment]
`./gradlew test` run all the tests of the project.

And I recommend you to look around.

**Comment by atulgpt** [comment]
Hi @BraisGabin the following TC is already passing. Should we close the issue?
```
@Test
fun `reports when property initiated with platform type `() {
    val code = """
        import java.net.URLEncoder

        class Person {
            val path = URLEncoder.encode("urlPath", "UTF-8")
        }
    """.trimIndent()
    assertThat(subject.compileAndLintWithContext(env, code)).hasSize(1)
}
```

**Comment by BraisGabin** [comment]
Closing then. Thanks!
